### PR TITLE
Fixes #11572 Use cfscrape to avoid Crunchyroll error 503

### LINKED ISSUE #11572: Crunchyroll Login page throwing 503
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like that [x])
- Use *Preview* tab to see how your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.12.22*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x ] I've **verified** and **I assure** that I'm running youtube-dl **2016.12.22**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add `-v` flag to **your command line** you run youtube-dl with, copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):
```
Youtube-DL>youtube-dl -i "http://www.crunchyroll.com/fairy-musketeers/episode-1-akazukin-finally-comes-510412" -u MyUserName -p MyPassword --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-i', 'http://www.crunchyroll.com/fairy-musketeers/episode-1-akazukin-finally-comes-510412', '-u', 'PRIVATE', '-p', 'PRIVATE', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2016.12.31
[debug] Python version 3.4.4 - Windows-7-6.1.7601-SP1
[debug] exe versions: ffmpeg N-62728-g6956b04, rtmpdump 2.5
[debug] Proxy map: {}
[Crunchyroll] Downloading login page
ERROR: Unable to download webpage: HTTP Error 503: Service Temporarily Unavailable (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest versi
on; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpa0c7afb5\build\youtube_dl\extractor\common.py", line 406, in _request_webpage
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpa0c7afb5\build\youtube_dl\YoutubeDL.py", line 2001, in urlopen
  File "C:\Python\Python34\lib\urllib\request.py", line 470, in open
  File "C:\Python\Python34\lib\urllib\request.py", line 580, in http_response
  File "C:\Python\Python34\lib\urllib\request.py", line 508, in error
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
  File "C:\Python\Python34\lib\urllib\request.py", line 588, in http_error_default

...

```

### Description of your *issue*, suggested solution and other information

If you provide login credentials for Crunchyroll, then the script shows a `HTTP Error 503: Service Temporarily Unavailable (caused by HTTPError())` error. But, if you remove the login credentials, it'll download the video/playlist just fine. Above, I have provided a log with the login credentials. I can open the login page just fine on all the browsers and I'm not behind any firewall. I have tried this with 2 different systems, in different location and I had the same result.

I checked the login page and it was just cloudflare browser check. There was no captcha to solve or anything, as mentioned in [this](https://github.com/rg3/youtube-dl/issues/11552) issue.

### PR DESCRIPTION
**Pull request by kidburglar**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Crunchyroll use not cloudfront when we login. To avoid a 503 error we need to insert the cookies that coudfront send.

I use cfscrape (https://github.com/Anorov/cloudflare-scrape) to grab the cookie that we need and insert them back. It is important that cfscrape have the same user-agent than youtube-dl, so it would be better not to hardcode it (I didn't find how I must do, so feel free to change it).

### CODE REVIEW

File: youtube_dl/extractor/crunchyroll.py [hunk 1]
```diff
@@ -43,6 +44,12 @@ def _login(self):
         if username is None:
             return
 
+        # Scrape cookie from cloudfront and insert them
+        scraper = cfscrape.create_scraper()
+        tokens = scraper.get_tokens(self._LOGIN_URL, 'Mozilla/5.0 (X11; Linux x86_64; rv:10.0) Gecko/20150101 Firefox/47.0 (Chrome)')
```
Comment (yan12125) [inline youtube_dl/extractor/crunchyroll.py]: Don't hardcode it. The default User-Agent can be found in ```std_headers``` in utils.py

File: youtube_dl/extractor/crunchyroll.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import json
 import base64
 import zlib
+import cfscrape
```
Comment (yan12125) [inline youtube_dl/extractor/crunchyroll.py]: Should handle ImportError and display proper messages to users. See https://github.com/rg3/youtube-dl/blob/master/youtube_dl/downloader/hls.py#L6 for an previous example.

**Comment by dstftw** [comment]
As far as I understand, MIT license is not compatible with Unlicense.

**Comment by yan12125** [comment]
If cfscrape is kept as an external dependency, I guess it's unrelated with youtube-dl's license? Just like standard Python modules, which is covered by Python license.

**Comment by kidburglar** [comment]
I have remove the hardcoded user-agent and use std_headers in place

File: youtube_dl/extractor/crunchyroll.py (hunk 2 above)
Comment (kidburglar) [inline youtube_dl/extractor/crunchyroll.py]: I create a check in _login that display a message if we use login function and that cscrape is not installed.
I don't find how completely quit the download after displaying the message, if you have some tips I will do the modifications.

**Comment by yan12125** [comment]
A question: does cloudflare protection occur in other pages or only login pages?

And cfscrape should be optional. The flow is like this:
```
if cfscrape_available:
    ...get_tokens_from_cfscrape...
else:
    ...warn_users_about_cloudflare...

...process_login...
```

**Comment by kidburglar** [comment]
Hello, sorry for the time I took I was a little busy these last days.

File: youtube_dl/extractor/crunchyroll.py [hunk 3]
```diff
@@ -39,6 +45,15 @@ class CrunchyrollBaseIE(InfoExtractor):
...
 
     def _login(self):
+        if cfscrape_available:
+            # Scrape cookie from cloudflare and insert them
+            scraper = cfscrape.create_scraper()
+            tokens = scraper.get_tokens(self._LOGIN_URL, std_headers['User-Agent'])
+            self._set_crunchyroll_cookie('cf_clearance', tokens[0]['cf_clearance'])
+            self._set_crunchyroll_cookie('__cfduid', tokens[0]['__cfduid'])
+        else
+            print 'cfscrape not found. Please install it if you want use login function for CrunchyRoll.'
```
Comment (yan12125) [inline youtube_dl/extractor/crunchyroll.py]: Use self._report_warning instead

**Comment by remitamine** [comment]
`AddAnimeIE` already contain some code to pass cloudfare challenge, i think it whould be better to use that code and improve it to handle the challenge in crunchyroll if it's different.

**Comment by kidburglar** [comment]
It look a little difficult to resolv the challenge with python (I don't know enough of it).
I see a function with firebug at the first answer for login page.

[function-login-page.txt](https://github.com/rg3/youtube-dl/files/721155/function-login-page.txt)

And it pass these parameters for the answer.

```
jschl_vc=xxx
pass=xxx
jschl_answer=1349
```
jschl_vc and pass seem ok with the code from AddAnimeIE

**Comment by yan12125** [comment]
> i think it whould be better to use that code and improve it to handle the challenge in crunchyroll if it's different.

> It look a little difficult to resolv the challenge with python (I don't know enough of it).

The code itself now looks fine. Before merging, I'd like to get this question answered. Is it infeasible to adopt AddAnimeIE's codes for Crunchyroll? @kidburglar

**Comment by kidburglar** [comment]
@yan12125 
I think it's possible to convert the javascript function in python but it look really complicated for me.
I find it proper and less breakable if we use cfscrape to interprete the javascript, resolve the challenge and use the cookies.

**Comment by yan12125** [comment]
My question is: is it inevitable to include Javascript interpretation to pass CloudFlare's DDoS check? If so using cfscrape is just fine.

By the way, would introducing such a big dependency affect Windows binaries? @dstftw

**Comment by kidburglar** [comment]
@yan12125 
What I understand of the javascript function is that we count "GQPqNQl.VEK" element and make multiple operation on it to finaly have a number. It's surely possible to do that in Python.

**Comment by yan12125** [comment]
> What I understand of the javascript function is that we count "GQPqNQl.VEK" element and make multiple operation on it to finaly have a number.

I'm afraid that Cloudflare can change this logic at any time. (For example use a different name each day) Anyway we can implement a first version and see how long will it work.

**Comment by kidburglar** [comment]
I just check and yes it seem change the name already.
It's for this reason I think cfscrape will work more longer, but we can maybe add a solution witthout cfscrape just in case some people cannot install it.

**Comment by Kcchouette** [comment]
@yan12125 and @remitamine when you speak about AddAnimeIE, you speak about [youtube_dl/extractor/addanime.py @ 15707c7](https://github.com/rg3/youtube-dl/blob/15707c7e024f1f29e7abd8ddaa362196ef2d4af6/youtube_dl/extractor/addanime.py) ?

**Comment by yan12125** [comment]
> when you speak about AddAnimeIE, you speak about youtube_dl/extractor/addanime.py @ 15707c7 ?

yep

**Comment by Kcchouette** [comment]
>  Is it infeasible to adopt AddAnimeIE's codes for Crunchyroll?

Can't you try to fix this issue using that, @yan12125 ?

**Comment by kidburglar** [comment]
@yan12125 
Any news about this ?
Because I not think I can do it without cfscrape, my python level is not good enough.

**Comment by yan12125** [comment]
Sorry for being so late on this. I forgot to say that cfscrape uses js2py, and the latter uses the dangerous ```exec()```. As youtube-dl is often used as a backend in online video downloading services, it's not a good idea to allow remote code execution without sandboxes.

**Comment by yan12125** [comment]
Technically those changes look good. But, as there are security issues, I marked it with "do-not-merge"

**Comment by kidburglar** [comment]
No problem.
Maybe you have some idea to improve the security or make a sandboxes ?

**Comment by yan12125** [comment]
There's going to be a pure-Python implementation for running Javascript at #11272, but I don't have time to check such a big change.

**Comment by yan12125** [comment]
As cfscrape drops js2py, we can go on.

Could you add a version check that ensures cfscrape >= 1.9.0 is used?

**Comment by yan12125** [comment]
Note to my self: read https://github.com/patriksimek/vm2/issues/32 carefully before merging anything

**Comment by kidburglar** [comment]
@yan12125 

I just checked on internet but I didn't find good way to looking about the version of a module without import it. Do you have an idea how to do that properly (or have an example) ?

**Comment by yan12125** [comment]
It's imported anyway, isn't it? Just make sure the version check occurs before actually using it.

**Comment by Kcchouette** [comment]
@yan12125 why merge this issue?
Isn't merging https://github.com/rg3/youtube-dl/pull/11272 a better solution for this issue and other related to JS-content website?

**Comment by yan12125** [comment]
@Kcchouette: I'm not sure what's going on in #11272. That PR contains only parsing routines, which is the first step of an interpreter, so it's not usable at the moment. Maybe the author is not working on it anymore.

**Comment by kidburglar** [comment]
@yan12125 It look that cfscrape doesn't have a __version__ method to give his method so the another way is to use pkg_ressources. It's a default package or I neet to do like cfscrape with a flag... ?

**Comment by yan12125** [comment]
> It look that cfscrape doesn't have a version method

A PR to cfscrape would be good

**Comment by kidburglar** [comment]
Hello @yan12125 

Sorry I didn't used youtube-dl for a very long time.
I cloned the repository again because I didn't used an apart branch for the PR so it was a real mess.

I'm not sure if it's more easier for you that I make a new PR, but I hope this time it will be accepted.
https://github.com/kidburglar/youtube-dl/commit/7d718bbeea2f9c846870bf072e89618aabaeadcb

Cheers.

**Comment by yan12125** [comment]
You'll need to create a new PR as GitHub can't merge pull requests from recreated repositories.

And you probably want to request a review from another active developer. As you may already know, I have been inactive in this project for quite a while.

**Pull request closed without merging**